### DISCUSSION TITLE
Feature Request: Add Steam as External OAuth Provider

### DISCUSSION THREAD

**Discussion by RiChrisMurphy** [supabase/supabase#4500] [Feature Requests]
Unsure of how difficult it would be to add, but I figured I'd ask. :)

**Comment by Isaiah-Hamilton** [comment]
I think this is a great Feature Request. You should start an issue in the [gotrue repository](https://github.com/supabase/gotrue) so people can further discuss it. And people from the auth team can see if this is doable.

**Comment by Johnrobmiller** [comment]
Even though I don't think there are a lot of Steam devs, the ones that are out there would LOVE an easy way to set up auth because right now it's a big pain

**Reply by gsabater** [reply]
There are dozens of us, dozens!

**Reply by Makosai** [reply]
I just found this after Googling for a way to handle Steam. So, yeah.

**Reply by dhunt-xsolla** [reply]
We exist!

**Reply by SteveChurch** [reply]
We do!

**Comment by vd3d** [comment]
I would like to promote my own suggestion https://github.com/orgs/supabase/discussions/30100 , because it include Steam and any other provider.

**Comment by Revadike** [comment]
[I attempted a PR for Steam provider support](https://github.com/supabase/auth/pull/1919)
However, I think the main problem is Supabase's reliability on user emails, so I am not sure if this will (fully) work.

**Comment by SasoriOlkof** [comment]
Hi, I'd be interested in implementing this and submitting a PR to `supabase/auth`.

I've reviewed the current external provider flow, and Steam appears to be implementable as a native provider despite using OpenID 2.0 rather than OAuth 2.0.

The proposed implementation would:

- authenticate the user through Steam OpenID 2.0;
- use the verified SteamID64 as the stable provider identity;
- support Steam identities without an email address;
- optionally enrich the identity with the user's Steam profile data;
- introduce no new external dependencies;
- include tests for the OpenID callback, invalid assertions, cancellation, PKCE, and callback replay.

A common reason for using Steam authentication is to associate users with a game they own. I am also considering an optional `RequiredAppID` setting that would allow self-hosted instances to reject accounts that do not own a configured Steam app. This would be disabled by default.

Before preparing the full PR, would the maintainers be open to native Steam support in `supabase/auth`? In particular, would you prefer the optional ownership check to be included in the initial PR or proposed separately?

**Reply by Revadike** [reply]
I suppose devs could just use an auth hook to make the ownership API call themselves, right?

**Reply by SasoriOlkof** [reply]
Yes, that is possible, and that is exactly the design question I am considering.

At the moment, I am preparing the implementation with ownership validation included as an optional feature. A self-hosted project could choose between:

- allowing any authenticated Steam user to sign in;
- requiring permanent ownership of a configured AppID;
- or also allowing access through Steam Families.

My concern is whether this should be proposed as part of the Steam provider itself or as a separate follow-up feature, so that native Steam authentication can be reviewed and accepted independently.

An auth hook could certainly perform the ownership check, but keeping it as an optional Supabase Auth configuration would allow developers to use Supabase as their only backend. They would not need to deploy and maintain a separate service just to verify ownership before granting access.

That said, I am happy to split the work if the maintainers would prefer the initial PR to cover only Steam authentication, with ownership validation discussed separately.

I have already started working on it here, although the branch is still a WIP:

https://github.com/SasoriOlkof/auth/tree/feat/steam-auth
